src/core/ddsi/src/ddsi_cdrstream.c

```diff
@@ -1284,6 +1374,15 @@ static const uint32_t *dds_stream_read_uni (dds_istream_t * __restrict is, char
 static inline const uint32_t *dds_stream_read_adr (uint32_t insn, dds_istream_t * __restrict is, char * __restrict data, const uint32_t * __restrict ops)
 {
   void *addr = data + ops[1];
+  if (op_type_external (insn) && DDS_OP_TYPE (insn) != DDS_OP_VAL_STR)
+  {
+    /* Allocate memory for @external struct member. This memory must be initialized
+        to 0, because the type may contain sequences that need to have 0 index/size */
+    uint32_t sz = get_adr_type_size (insn, ops);
+    *((char **) addr) = ddsrt_calloc (1, sz);
```

Comment:
Doesn't this lead to a memory leak if one has a loop:
```C
T *x = 0;
dds_sample_info_t si;
while((n=dds_take(rd, &x, si, 1, 1))>0) { ... }
...
if (x) dds_return_loan(rd,&x,1);
```

I'm commenting here because the union case is already special-cased to deal with changing discriminants. (See `dds_stream_read_sample`.)

Reply:
Thanks for finding this! I've fixed it in 198041c. The size of the allocated buffer for external members is a constant value independent of the data, so I think it can be safely re-used over multiple reads/takes (similar to string and sequence buffers).